Tests/test-dataframe.lisp

```diff
@@ -0,0 +1,554 @@
+    (signals <invalid-slice> (sel df row-slice* col-slice*))))
+
+(def-test ref ()
+  (for-all (((df row-spec col-index)
+             (lambda ()
+               (let ((df (funcall
+                          (gen-data-frame (funcall (gen-integer :min 1 :max +max-rows+))
+                                          (funcall (gen-integer :min 1 :max +max-cols+))))))
+                 (list df
+                       (funcall (gen-valid-row-spec df))
+                       (funcall (gen-valid-col-index df)))))))
+    (finishes (ref df row-spec col-index))))
+
+(def-test ref.invalid ()
+  (for-all (((df row-spec col-index row-spec* col-index*)
+             (lambda ()
+               (let ((df (funcall
+                          (gen-data-frame (funcall (gen-integer :min 1 :max +max-rows+))
+                                          (funcall (gen-integer :min 1 :max +max-cols+))))))
+                 (list df
+                       (funcall (gen-valid-row-spec df))
+                       (funcall (gen-valid-col-index df))
+                       (funcall (gen-invalid-row-spec df))
+                       (funcall (gen-invalid-col-index df)))))))
+    (signals <invalid-index> (ref df row-spec col-index*))
+    (signals <invalid-index> (ref df row-spec* col-index))
+    (signals <invalid-index> (ref df row-spec* col-index*))))
+
+(def-test map-data-frame ()
+  (for-all (((df row-slice col-slice function)
+             (lambda ()
+               (let ((df (funcall
+                          (gen-data-frame (funcall (gen-integer :min 1 :max +max-rows+))
+                                          (funcall (gen-integer :min 1 :max +max-cols+))))))
+                 (list df
+                       (funcall (gen-valid-row-slice df))
+                       (funcall (gen-valid-col-slice df))
+                       (funcall (gen-valid-function 'map)))))))
+    (finishes (map-data-frame df row-slice col-slice function))))
+
+(def-test map-data-frame.invalid ()
+  (for-all (((df row-slice col-slice row-slice* col-slice*)
+             (lambda ()
+               (let ((df (funcall
+                          (gen-data-frame (funcall (gen-integer :min 1 :max +max-rows+))
+                                          (funcall (gen-integer :min 1 :max +max-cols+))))))
+                 (list df
+                       (funcall (gen-valid-row-slice df))
+                       (funcall (gen-valid-col-slice df))
+                       (funcall (gen-invalid-row-slice df))
+                       (funcall (gen-invalid-col-slice df))))))
+            (function (gen-valid-function 'map))
+            (function* (gen-invalid-function 'map)
+                       (if (or (and (arrayp row-slice)
+                                    (zerop (array-total-size row-slice)))
+                               (and (arrayp col-slice)
+                                    (zerop (array-total-size col-slice))))
+                           nil
+                           t)))
```

Comment:
replace if with just a test

`(not (or (and ) (and ))`

Reply:
Done.